pir/cpp/client.cpp

```diff
@@ -41,69 +46,58 @@ StatusOr<std::unique_ptr<PIRClient>> PIRClient::Create(
 }
 
 StatusOr<PIRPayload> PIRClient::CreateRequest(std::size_t index) const {
-  if (index >= this->DBSize()) {
+  const auto poly_modulus_degree =
+      context_->Parameters()->GetEncryptionParams().poly_modulus_degree();
+  if (index >= DBSize()) {
     return InvalidArgumentError("invalid index");
   }
-  std::vector<std::int64_t> request(DBSize(), 0);
-  request[index] = 1;
-
-  return this->encryptRequestBuffer(request);
+  if (index >= poly_modulus_degree) {
+    // Not yet implemented
+    return InvalidArgumentError("More than 1 CT needed for selection vector");
+  }
+  Plaintext pt(poly_modulus_degree);
+  pt.set_zero();
+  pt[index] = 1;
+  std::cout << "Query PT: " << pt.to_string() << std::endl;
+
+  vector<Ciphertext> query(1);
+  GaloisKeys gal_keys;
+  try {
+    encryptor_->encrypt(pt, query[0]);
+    gal_keys =
+        keygen_->galois_keys_local(generate_galois_elts(poly_modulus_degree));
+  } catch (const std::exception& e) {
+    return InternalError(e.what());
+  }
+  return PIRPayload::Load(query, gal_keys);
 }
 
 StatusOr<int64_t> PIRClient::ProcessResponse(const PIRPayload& response) const {
-  ASSIGN_OR_RETURN(auto decrypted, this->decryptResponseBuffer(response));
-  std::map<uint64_t, int64_t> result;
-
-  for (size_t idx = 0; idx < decrypted.size(); ++idx) {
-    if (decrypted[idx] != 0) {
-      return decrypted[idx];
-    }
+  if (response.Get().size() != 1) {
+    return InvalidArgumentError("Number of ciphertexts in response must be 1");
   }
-  return InvalidArgumentError("invalid server response");
-}
-
-StatusOr<PIRPayload> PIRClient::encryptRequestBuffer(
-    const std::vector<int64_t>& in) const {
-  std::vector<seal::Ciphertext> ciphertexts(in.size());
-
-  for (size_t idx = 0; idx < in.size(); ++idx) {
-    seal::Ciphertext ciphertext(context_->SEALContext());
-
-    seal::Plaintext plaintext;
-
-    try {
-      context_->Encoder()->encode(in[idx], plaintext);
-    } catch (const std::exception& e) {
-      return InvalidArgumentError(e.what());
-    }
-
-    try {
-      encryptor_->encrypt(plaintext, ciphertext);
-    } catch (const std::exception& e) {
-      return InvalidArgumentError(e.what());
-    }
-    ciphertexts[idx] = ciphertext;
+  const uint32_t m = next_power_two(DBSize());
+
+  seal::Plaintext plaintext;
+  try {
+    decryptor_->decrypt(response.Get()[0], plaintext);
+    // have to divide the integer result by the the next power of 2 greater than
```

Comment:
Note that this can be done during CreateRequest() instead by doing pt[index] = 1/m; At the end of the expansion, the result would therefore be = 1. Also, oblivious expansion requires the plaintext space to be odd, this should probably be emphasized somewhere (if it is not done yet).

Reply:
Oh very good call! I've updated CreateRequest to do that. Thanks!